### PR TITLE
feat(planner_manager): limit iteration number by parameter

### PR DESCRIPTION
**Pull request by satoshi-ota**
### <samp>🤖 Generated by Copilot at aa9464c</samp>

This pull request adds a new parameter `max_itration_num` to the behavior path planner node and its components. The parameter controls the maximum number of iterations for the planner manager to run the behavior modules and select the best output. This improves the performance and stability of the path planning process.

Please review following PR at first:
https://github.com/autowarefoundation/autoware_launch/pull/645

## Tests performed

- [ ] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/7bbe67d4-40d5-53ef-87b5-eb0c8c96f7c7?project_id=prd_jt)

## Effects on system behavior

Nothing.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/config/behavior_path_planner.param.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /**:
   ros__parameters:
     verbose: false
+    max_itration_num: 100
```
Comment (rej55) [inline planning/behavior_path_planner/config/behavior_path_planner.param.yaml:4]:
```suggestion
    max_iteration_num: 100
```

File: planning/behavior_path_planner/include/behavior_path_planner/parameters.hpp [hunk 2]
```diff
@@ -77,6 +77,7 @@ struct LateralAccelerationMap
 struct BehaviorPathPlannerParameters
 {
   bool verbose;
+  size_t max_itration_num{100};
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/parameters.hpp:80]:
```suggestion
  size_t max_iteration_num{100};
```

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp [hunk 3]
```diff
@@ -97,7 +97,7 @@ struct SceneModuleStatus
 class PlannerManager
 {
 public:
-  PlannerManager(rclcpp::Node & node, const bool verbose);
+  PlannerManager(rclcpp::Node & node, const size_t max_itration_num, const bool verbose);
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:100]:
```suggestion
  PlannerManager(rclcpp::Node & node, const size_t max_iteration_num, const bool verbose);
```

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp [hunk 4]
```diff
@@ -446,6 +446,8 @@ class PlannerManager
 
   mutable std::shared_ptr<SceneModuleVisitor> debug_msg_ptr_;
 
+  size_t max_itration_num_{100};
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp:449]:
```suggestion
  size_t max_iteration_num_{100};
```

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp [hunk 5]
```diff
@@ -132,7 +132,7 @@ BehaviorPathPlannerNode::BehaviorPathPlannerNode(const rclcpp::NodeOptions & nod
     const std::lock_guard<std::mutex> lock(mutex_manager_);  // for planner_manager_
 
     const auto & p = planner_data_->parameters;
-    planner_manager_ = std::make_shared<PlannerManager>(*this, p.verbose);
+    planner_manager_ = std::make_shared<PlannerManager>(*this, p.max_itration_num, p.verbose);
```
Comment (rej55) [inline planning/behavior_path_planner/src/behavior_path_planner_node.cpp:135]:
```suggestion
    planner_manager_ = std::make_shared<PlannerManager>(*this, p.max_iteration_num, p.verbose);
```

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp [hunk 6]
```diff
@@ -272,6 +272,7 @@ BehaviorPathPlannerParameters BehaviorPathPlannerNode::getCommonParam()
   BehaviorPathPlannerParameters p{};
 
   p.verbose = declare_parameter<bool>("verbose");
+  p.max_itration_num = declare_parameter<int>("max_itration_num");
```
Comment (rej55) [inline planning/behavior_path_planner/src/behavior_path_planner_node.cpp:275]:
```suggestion
  p.max_iteration_num = declare_parameter<int>("max_iteration_num");
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 7]
```diff
@@ -29,9 +29,11 @@
 
 namespace behavior_path_planner
 {
-PlannerManager::PlannerManager(rclcpp::Node & node, const bool verbose)
+PlannerManager::PlannerManager(
+  rclcpp::Node & node, const size_t max_itration_num, const bool verbose)
```
Comment (rej55) [inline planning/behavior_path_planner/src/planner_manager.cpp:33]:
```suggestion
  rclcpp::Node & node, const size_t max_iteration_num, const bool verbose)
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 8]
```diff
@@ -29,9 +29,11 @@
...
 : logger_(node.get_logger().get_child("planner_manager")),
   clock_(*node.get_clock()),
+  max_itration_num_{max_itration_num},
```
Comment (rej55) [inline planning/behavior_path_planner/src/planner_manager.cpp:36]:
```suggestion
  max_iteration_num_{max_iteration_num},
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 9]
```diff
@@ -128,8 +130,17 @@ BehaviorModuleOutput PlannerManager::run(const std::shared_ptr<PlannerData> & da
       addApprovedModule(highest_priority_module);
       clearCandidateModules();
       debug_info_.emplace_back(highest_priority_module, Action::ADD, "To Approval");
+
+      if (itr_num >= max_itration_num_) {
```
Comment (rej55) [inline planning/behavior_path_planner/src/planner_manager.cpp:134]:
```suggestion
      if (itr_num >= max_iteration_num_) {
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 10]
```diff
@@ -128,8 +130,17 @@ BehaviorModuleOutput PlannerManager::run(const std::shared_ptr<PlannerData> & da
...
+        RCLCPP_WARN_THROTTLE(
+          logger_, clock_, 1000, "Rearch iteration limit (max: %ld). Output current result.",
+          max_itration_num_);
```
Comment (rej55) [inline planning/behavior_path_planner/src/planner_manager.cpp:137]:
```suggestion
          max_iteration_num_);
```

File: planning/behavior_path_planner/config/behavior_path_planner.param.yaml (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/config/behavior_path_planner.param.yaml]: Fixed.

File: planning/behavior_path_planner/include/behavior_path_planner/parameters.hpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/parameters.hpp]: Fixed.

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp (hunk 3 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp]: Fixed.

File: planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp (hunk 4 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/planner_manager.hpp]: Fixed.

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp (hunk 5 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp]: Fixed.

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp (hunk 6 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp]: Fixed.

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 7 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed.

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 9 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed.

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 10 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed.

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 11]
```diff
@@ -128,8 +130,17 @@ BehaviorModuleOutput PlannerManager::run(const std::shared_ptr<PlannerData> & da
       addApprovedModule(highest_priority_module);
       clearCandidateModules();
       debug_info_.emplace_back(highest_priority_module, Action::ADD, "To Approval");
+
+      if (itr_num >= max_iteration_num_) {
+        RCLCPP_WARN_THROTTLE(
+          logger_, clock_, 1000, "Rearch iteration limit (max: %ld). Output current result.",
```
Comment (rej55) [inline planning/behavior_path_planner/src/planner_manager.cpp:136]:
```suggestion
          logger_, clock_, 1000, "Reach iteration limit (max: %ld). Output current result.",
```

Comment (satoshi-ota) [reply planning/behavior_path_planner/src/planner_manager.cpp]: Fixed. :bow:

**Review by rej55 — APPROVED** [review]

**Pull request merged**